server/routerlicious/packages/services-telemetry/src/lumber.ts

```diff
@@ -0,0 +1,140 @@
+        return this._properties;
+    }
+
+    public get durationInMs(): number | undefined {
+        if (this.type === LumberType.Log) {
+            return undefined;
+        }
+        return this._durationInMs;
+    }
+
+    public get successful(): boolean | undefined {
+        if (this.type === LumberType.Log) {
+            return undefined;
+        }
+        return this._successful;
+    }
+
+    public get message(): string | undefined {
+        return this._message;
+    }
+
+    public get statusCode(): string | undefined {
+        return this._statusCode;
+    }
+
+    public get exception(): Error | undefined {
+        return this._exception;
+    }
+
+    public get logLevel(): LogLevel | undefined {
+        return this._logLevel;
+    }
+
+    constructor(
+        public readonly eventName: T,
+        public readonly type: LumberType,
+        private readonly _engineList: ILumberjackEngine[],
+        private readonly _schemaValidator?: ILumberjackSchemaValidator,
+        properties?: Map<string, any> | Record<string, any>) {
+            if (properties) {
+                this.addProperties(properties);
+            }
+        }
+
+    public addProperty(key: string, value: any): this {
+        this._properties.set(key, value);
+        return this;
+    }
+
+    public addProperties(properties: Map<string, any> | Record<string, any>): this {
+        if (properties instanceof Map) {
+            if (this._properties.size === 0) {
+                this._properties = properties;
+            } else {
+                properties.forEach((value: any, key: string) => { this.addProperty(key, value); });
+            }
+        } else {
+            Object.entries(properties).forEach((entry) => {
+                const [key, value] = entry;
+                this.addProperty(key,value);
+            });
+        }
+        return this;
+    }
+
+    public success(
+        message: string,
+        statusCode: number | string | undefined,
+        logLevel: LogLevel = LogLevel.Info) {
+        this.emit(message, statusCode, logLevel, true);
+    }
+
+    public error(
+        message: string,
+        statusCode: number | string | undefined,
+        exception?: Error | undefined,
+        logLevel: LogLevel = LogLevel.Error) {
+        this.emit(message, statusCode, logLevel, false, exception);
+    }
+
+    private emit(
+        message: string,
+        statusCode: number | string | undefined,
+        logLevel: LogLevel,
+        successful: boolean,
+        exception?: Error) {
+        if (this._completed) {
+            throw new Error(
+                `Trying to complete a Lumber telemetry operation ${this.eventName} that has alredy been completed.`);
+        }
+
+        if (this._schemaValidator) {
+            const validation = this._schemaValidator.validate(this.properties);
+
+            if (!validation.validationPassed) {
+                throw new Error(
```

Comment:
I had a discussion with @TriDucNgo and @znewton and we realized that it can be dangerous to throw an error if schema validation fails. Since this would happen during runtime, it could cause an Error to happen just because of the schema. If code is checked in and not validated/run locally, it could potentially add telemetry that would cause this error in Production. Instead, we will adopt the more conservative approach of emitting a log/message indicating the schema error.